File: 2019/python/day2.py
def run(pgm, noun, verb):
    p = list(pgm)
    p[1] = noun
    p[2] = verb
    ip = 0
    while p[ip] != 99:
        op, a, b, r = p[ip : ip+4]
        if op == 1:
            p[r] = p[a] + p[b]
        elif op == 2:
            p[r] = p[a] * p[b]
        ip += 4

    return p[0]

with open('../day_2.txt', 'r') as f:
    pgm = [int(s) for s in f.read().rstrip().split(',')]

print('part 1:', run(pgm, 12, 2))

r = next(n*100+v for n in range(100) for v in range(100) if run(pgm, n, v) == 19690720)
print('part 2:', r)

File: 2019/go/day2.go
package main

import (
	"AdventOfCode/intcode"
	"AdventOfCode/util"
	"fmt"
)

func part1(pgm string) int {
	computer, _ := intcode.LoadProgram(pgm)
	computer.Mem[1] = 12
	computer.Mem[2] = 2
	computer.Run()
	return computer.Mem[0]
}

func part2(pgm string, expected int) int {
	for noun := 0; noun < 100; noun++ {
		for verb := 0; verb < 100; verb++ {
			computer, _ := intcode.LoadProgram(pgm)
			computer.Mem[1] = noun
			computer.Mem[2] = verb
			computer.Run()
			if computer.Mem[0] == expected {
				return noun*100 + verb
			}
		}
	}
	return -1
}

func main() {
	input, _ := util.ReadInput("../day_2.txt")
	pgm := input[0]
	fmt.Printf("part 1: %d\n", part1(pgm))
	fmt.Printf("part 2: %d\n", part2(pgm, 19690720))
}
